### ISSUE TITLE
%var% is not replaced by blank when is not defined in the config file

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#404]
In GitLab by @pablohe on May 31, 2019, 11:49

The problem was detected in `3.12.1b0` version

When a variable used in the templates (e.g. `%var%`) 
is not defined in config files the final value is not changed i.e. autosubmit left `%var%` value

The behavior should be replace by blank
 
@etiennesky @mcastril @ jberlin

**Comment by kinow** [comment]
In GitLab by @pablohe on May 31, 2019, 11:50

maybe related with #401

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 4, 2019, 16:54

mentioned in commit 2751fd0f9cfb6d1d3add1e9bc0964df50f9bdac5

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 4, 2019, 17:01

Will close this because is the same error than in #401 and similar to #400 ( updated info in #400)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 4, 2019, 17:01

closed

**Comment by kinow** [comment]
In GitLab by @pablohe on Jun 11, 2019, 13:39

mentioned in issue auto-ecearth3#938

**Comment by kinow** [comment]
In GitLab by @pablohe on Jun 13, 2019, 11:39

I'll reopen it because is more general (now the problem is in SYNCHRONIZE Job)

I run it with inspect option
 
In the cmd command I can found %var% inside 

you can check the `t04r` tmp folder

**Comment by kinow** [comment]
In GitLab by @pablohe on Jun 13, 2019, 11:39

reopened

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 13, 2019, 11:56

hi @pablohe ,

not sure how to reproduce this I have this config and works for me in 3.12.0b:

```
[transfer_node]
TYPE = ps
HOST = dt
USER = bsc32070
PROJECT = bsc32
ADD_PROJECT_TO_HOST = false
SCRATCH_DIR = /gpfs/scratch
```

jobs:

```
[SYNCHRONIZE]
FILE = common.synchronize.sh
PLATFORM = transfer_node
#DEPENDENCIES = SYNCHRONIZE-1  # commented to test member
#RUNNING = member # same
CHECK = false
```

```
cat ../tmp/

#
# Architecture
#

HPCARCH=transfer_node
PROJDIR=/esarchive/autosubmit/a1wy/proj/debug
SCRATCH_DIR=/gpfs/scratch
HPCPROJ=bsc32
HPCUSER=bsc32070
EXPID=a1wy
PROJDEST=debug
ROOTDIR=/gpfs/scratch/bsc32/bsc32070/a1wy
HPCHOST=dt
UPDATE_MODEL=
DTHOST=
USE_DT_MACHINE=
DT_HOST=
DT_USER=

```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 13, 2019, 12:30

Hi @pablohe

This is so weird, sometimes it works without change anything and sometimes not I guess it will take more time

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 14, 2019, 13:52

Hi @pablohe,

It should works on 3.12.0b that is the one you use now as weekly test ( no?)together with #416, #409.

As you can see in #416, now there is a change in `check`:

in jobs.conf you can select three options for `check`:

* `Check=True` enable logs and debug logs for check_scripts
* `Check=False` disable logs
* `Check=On_submission` disable the initial check for scripts that will be generated in run-time.

Limitations of `On_submission`, 

* logs are disabled for it for now as discussed with @mcastril.
* `autosubmit inspect <expid>` is unable to `check=on_submission` scripts since it doesn't submit anything so the `on_submission` scripts won't generate

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 14, 2019, 14:44

Thanks @dbeltrankyl 

> `Check=False` disable logs 

This means that the user will see the warning about the script but not a warning per variable right?

> `Check=On_submission` disable the initial check for scripts that will be generated in run-time. 

With this you can disable all warnings until submit time arrives. Then you will see a general warning for the script but not the variables right?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 14, 2019, 14:56

Hi ,

`check = False` output will be similar to 
```
checking scripts...
[WARNING] Template SYNCHRONIZE will be checked without logs
[WARNING] Template INI will be checked without logs
[WARNING] Template CMOROCE will be checked without logs
[WARNING] Template CMORATM will be checked without logs
[WARNING] Template POST will be checked without logs
[WARNING] Template CLEAN will be checked without logs
[WARNING] Template CLEAN_MEMBER will be checked without logs
[WARNING] Template TRANSFER will be checked without logs
[WARNING] Template PERFORMANCE_METRICS will be checked without logs
[WARNING] Template TRANSFER_MEMBER will be checked without logs

And the last output:

Scripts OK 
or             
Log.warning("Scripts check failed")
Log.user_warning("Running after failed scripts check is at your own risk!")
```

`Check=On_submission` on Sim job per example, 

Won't show and do anything until running time for the job that has this option set

Once it is in running time, if the script has created successfully by another job, it will check blank values and placeholders but it won't show anything although, I can add `[WARNING] Template SIM will be checked without logs ` if you desire, However it would be repeated by `C*M*D (chunks,member, dates)`.

If the template script .sh doesn't exist AS will crash since there was something wrong in a completed job that didn't generate the SIM script and autosubmit can't know it.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 14, 2019, 15:04

Thanks @dbeltrankyl . I think we can leave as it is now and think of a solution for only printing one warning when the SIM job is actually submitted (for every SIM job).

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2019, 14:43

Hi @dbeltrankyl ,

I am having a problem testing the last version (3.12.1b). When starting an Auto-EC-Earth experiment I have:

```
[CRITICAL] Unhandled exception on Autosubmit: [Errno 2] No such file or directory: '/esarchive/autosubmit/a21x/proj/auto-ecearth3/sources/runtime/autosubmit/ece-esm.sh'
Traceback (most recent call last):
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/autosubmit.py", line 418, in parse_args
    return Autosubmit.run_experiment(args.expid, args.notransitive)
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/autosubmit.py", line 956, in run_experiment
    job_list.check_scripts(as_conf)
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/job/job_list.py", line 935, in check_scripts
    if not job.check_script(as_conf, self.parameters,show_logs):
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/job/job.py", line 832, in check_script
    template_content = self.update_content(as_conf)
  File "/shared/earth/software/autosubmit/3.12.1b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.1b0-py2.7.egg/autosubmit/job/job.py", line 705, in update_content
    template_file = open(os.path.join(as_conf.get_project_dir(), self.file), 'r')
IOError: [Errno 2] No such file or directory: '/esarchive/autosubmit/a21x/proj/auto-ecearth3/sources/runtime/autosubmit/ece-esm.sh'
```

As you know, this file is generated in the first job, so it doesn't exist before running the experiment. If I use CHECK = On_Submission (as we do most of the times) it works, but it doesn't work if I set CHECK = False. I think I understand why, taking into account our discussions about how the CHECK works, but it may be confusing for users.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 3, 2019, 14:53

Hi,

This is the normal behavior now,

The problem before the update was that if you were putting check = False it would work for that SIM job but It wouldn't replace blank values and placeholders that I think it should be checked at all cases.

So now, what check does is to Enable/disable logs and with on_submission it will check the script but in running time. But the user can now ignore this variable and only be worried about check=on_submission and additionally select if they want to see logs maybe I could change true/false to:

```
check unset == non logs 
Check = LOGS active them.
Check = On_submission current behaviour.
```

However, I'm unsure if it would be more confusing than now.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 3, 2019, 15:08

The problem is that I also tried to comment the CHECK variable in case I could unset it, but it was also failing. So the only way I had to run this experiment was to put CHECK = On_submission.

I think in the future we should try to split the logic of the Autosubmit parser from the variable warnings. For the user side it is not clear that replacing involves checking.

The requirements should be:

- Blank values and placeholders should always be substituted (whenever the file exist). If the template exists on `Autosubmit run` can do it there, if not it always can try doing it at the submission time.

- The `CHECK` variable should only be used for logs (errors): FALSE/TRUE to disable/enable, and On_submission to perform the check only on submission time.

So basically it would be as it is now, but setting `CHECK = FALSE` would not turn off variable replacing, and should not prompt errors about missing templates. In any case this is not a priority because it is not preventing anything to run, it can be done in some months.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 5, 2019, 11:15

Closing this issue, opening a new one in #425 with the last mentioned by @mcastril.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 5, 2019, 11:17

closed

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 18, 2019, 10:11

mentioned in issue auto-ecearth3#1075
